README.md
[![Build Status](https://app.travis-ci.com/fivdi/mcp-spi-adc.svg?branch=master)](https://app.travis-ci.com/github/fivdi/mcp-spi-adc)
[![npm Version](http://img.shields.io/npm/v/mcp-spi-adc.svg)](https://www.npmjs.com/package/mcp-spi-adc)
[![Downloads Per Month](http://img.shields.io/npm/dm/mcp-spi-adc.svg)](https://www.npmjs.com/package/mcp-spi-adc)

# mcp-spi-adc

MCP3002/4/8, MCP3201/2/4/8 and MCP3304 SPI analog to digital conversion with
**Node.js** on Linux boards like the Raspberry Pi or BeagleBone.

mcp-spi-adc supports Node.js versions 10, 12, 14, 15 and 16.

## Contents

 * [Installation](#installation)
 * [Usage](#usage)
 * [Supported Devices](#supported-devices)
 * [API Documentation](#api-documentation)

## Installation

```
npm install mcp-spi-adc
```

In order to use mcp-spi-adc SPI must be enabled. How SPI is enabled varies
from board to board. For example, on the Raspberry Pi the raspi-config tool
can be used enable SPI. On the BeagleBone Black the config-pin utility can be
used.

## Usage

Determine the temperature using a TMP36 analog temperature sensor wired to
channel 5 on an MCP3008 SPI A/D converter.

<img src="https://raw.githubusercontent.com/fivdi/mcp-spi-adc/master/example/pi-mcp3008-tmp36.png">

```js
const mcpadc = require('mcp-spi-adc');

const tempSensor = mcpadc.open(5, {speedHz: 20000}, err => {
  if (err) throw err;

  setInterval(_ => {
    tempSensor.read((err, reading) => {
      if (err) throw err;

      console.log((reading.value * 3.3 - 0.5) * 100);
    });
  }, 1000);
});
```

Note how the optional configuration option speedHz is used to configure the
SPI clock frequency in Hertz for reading the value from the TMP36 temperature
sensor. The default SPI clock frequency for the MCP3008 is 1350000Hz but
lowering it to 20000Hz gives a more acurate temperature reading. In general,
it's not necessary to lower the clock speed to read a value.

The default clock speed of 1350000Hz for the MCP3008 is derived from the
[MCP3008 datasheet](https://cdn-shop.adafruit.com/datasheets/MCP3008.pdf).
The maximum sampling rate at VDD = 2.7V is 75 ksps and each sample requires
an 18-bit transfer. 75000 x 18 = 1350000. 1350000Hz is a conservative frequency
in the above circuit as VDD is 3.3V.

## Supported Devices

Device | Channels | Channel Numbers | Default Clock Frequency | Resolution | Raw Value Range
:---: | ---: | ---: | ---: | ---: | ---:
MCP3002 | 2 | 0-1 | 1200000Hz | 10-bit | 0-1023
MCP3004 | 4 | 0-3 | 1350000Hz | 10-bit | 0-1023
MCP3008 | 8 | 0-7 | 1350000Hz | 10-bit | 0-1023
MCP3201 | 1 | 0 | 800000Hz | 12-bit | 0-4095
MCP3202 | 2 | 0-1 | 900000Hz | 12-bit | 0-4095
MCP3204 | 4 | 0-3 | 1000000Hz | 12-bit | 0-4095
MCP3208 | 8 | 0-7 | 1000000Hz | 12-bit | 0-4095
MCP3304 | 8 | 0-7 | 1050000Hz | 13-bit | 0-4095

## API Documentation

All methods are asynchronous and take a completion callback as their last
argument. The arguments passed to the completion callback depend on the
method, but the first argument is always reserved for an exception. If the
operation was completed successfully, then the first argument will be null
or undefined.

### Functions

- [openMcp3002(channel[, options], cb)](#openmcp3002channel-options-cb)
- [openMcp3004(channel[, options], cb)](#openmcp3004channel-options-cb)
- [openMcp3008(channel[, options], cb)](#openmcp3008channel-options-cb)
- [openMcp3201(channel[, options], cb)](#openmcp3201channel-options-cb)
- [openMcp3202(channel[, options], cb)](#openmcp3202channel-options-cb)
- [openMcp3204(channel[, options], cb)](#openmcp3204channel-options-cb)
- [openMcp3208(channel[, options], cb)](#openmcp3208channel-options-cb)
- [openMcp3304(channel[, options], cb)](#openmcp3304channel-options-cb)
- [open(channel[, options], cb) - alias for openMcp3008(channel[, options], cb)](#openchannel-options-cb---alias-for-openmcp3008channel-options-cb)

### Class AdcChannel

- [adcChannel.read(cb)](#adcchannelreadcb)
- [adcChannel.close(cb)](#adcchannelclosecb)

### openMcp3002(channel[, options], cb)
### openMcp3004(channel[, options], cb)
### openMcp3008(channel[, options], cb)
### openMcp3201(channel[, options], cb)
### openMcp3202(channel[, options], cb)
### openMcp3204(channel[, options], cb)
### openMcp3208(channel[, options], cb)
### openMcp3304(channel[, options], cb)
### open(channel[, options], cb) - alias for openMcp3008(channel[, options], cb)
- channel - the number of the channel to open, see channel numbers in
[supported devices](#supported-devices)
- options - an optional object specifying channel configuration options
- cb - completion callback

Asynchronous open. Returns a new AdcChannel object. The completion callback
gets one argument (err). The AdcChannel object returned should not be used
before the completion callback is called.

The following channel configuration options are supported:

- busNumber - the SPI bus number, 0 for `/dev/spidev0.n`,
1 for `/dev/spidev1.n`, ..., default 0
- deviceNumber - the SPI device number, 0 for `/dev/spidevn.0`,
1 for `/dev/spidevn.1`, ..., default 0
- speedHz - a number representing the SPI clock frequency for reading from the
channel in Hertz, see default clock frequency in
[supported devices](#supported-devices)

### adcChannel.read(cb)
- cb - completion callback

Asynchronous read. The completion callback gets two arguments (err,
reading). The reading argument is an object with the following properties:

- value - the value read from the channel scaled to a value between 0 and 1
- rawValue - the value read from the channel, see raw value range in
[supported devices](#supported-devices)

Returns this.

### adcChannel.close(cb)
- cb - completion callback

Asynchronous close. Frees system resources used by this instance. The
completion callback gets one argument (err). Returns null.



History.md
3.2.0 / Oct 17 2021
===================

 * add mcp3201 support

3.1.1 / Apr 30 2021
===================

 * drop support for node.js 8 and 13, add support for node.js 15 and 16
 * update dependencies

3.1.0 / Apr 25 2020
===================

 * update dependencies (spi-device v3.1.0, jshint v2.11.0)
 * drop support for node.js 6, add support for node.js 14

3.0.0 / Sep 22 2019
===================

 * update dependencies (spi-device v3.0.0)
 * drop support for node.js v4

2.0.7 / Sep 07 2019
===================

 * update dependencies (spi-device v2.0.9)

2.0.6 / Jun 16 2019
===================

  * remove node 11 from build
  * update dependencies
  * update npm keywords

2.0.5 / Mar 15 2019
===================

 * update dependencies (spi-device v2.0.7)
 * document node 12 support
 * lint with jshint
 * add travis build
 * add badges

2.0.4 / Oct 02 2018
===================

 * update dependencies (spi-device v2.0.4)

2.0.3 / Jul 28 2018
===================

 * update dependencies (spi-device v2.0.3)
 * code style

2.0.2 / Apr 15 2018
===================

 * update dependencies (spi-device v2.0.2, lodash.clone v4.5.0)
 * code style
 * add test to call open, read and close for all supported chips

2.0.1 / Apr 07 2018
===================

 * update dependencies (spi-device v2.0.1)

2.0.0 / Feb 25 2018
===================

 * document node 9 support
 * update dependencies (spi-device v2.0.0)
 * drop support for node.js v0.10, v0.12, v5 and v7

1.0.0 / Oct 14 2017
===================

 * update dependencies (spi-device v1.0.0)
 * document supported node versions

0.4.0 / Jan 05 2017
===================

 * MCP3304 support

0.3.1 / Oct 03 2016
===================

 * bugfix: the mcp3002 has 2 channels not 8

0.3.0 / Oct 02 2016
===================

 * MCP3002 support
 * MCP3202 support

0.2.1 / Oct 02 2016
===================

 * documentation

0.2.0 / Oct 02 2016
===================

 * MCP3004 support
 * MCP3204 support

0.1.0 / Oct 02 2016
===================

 * MCP3208 support

0.0.1 / May 31 2016
===================

  * documentation

0.0.0 / May 28 2016
===================

  * initial release



test/channel-numbers.js
'use strict';

const mcpadc = require('../');
const assert = require('assert');

const channels = [
  // ok
  0,
  7,

  // errors
  -1,
  8,
  undefined,
  null,
  'Hello, World!',
  '',
  '4',
  [],
  {},
  _ => {},
  true,
  false,
  3.142,
  NaN,
  Infinity
];

channels.forEach(channel => {
  try {
    const sensor = mcpadc.open(channel, {}, err => {
      assert(!err, 'can\'t open sensor');
      assert(channel >= 0 && channel <= 7 , 'invalid channel opened');

      console.log('  channel: ' + channel);

      sensor.close(err => {
        assert(!err, 'can\'t close sensor');
      });
    });
  } catch (e) {
    assert(channel !== 0 && channel !== 7, 'valid channel not opened');
  }
});



test/open-read-close.js
'use strict';

const mcpadc = require('../');
const assert = require('assert');

[0, 1, 2, 3, 4].forEach(channelNumber => {
  const sensor = mcpadc.open(channelNumber, {}, err => {
    assert(!err, 'can\'t open sensor');
    sensor.read((err, reading) => {
      assert(!err, 'can\'t read sensor');
      console.log('  ch' + channelNumber + ': ' +
        reading.rawValue + ' / ' + reading.value);
      sensor.close(err => {
        assert(!err, 'can\'t close sensor');
      });
    });
  });
});



test/stress/two-channels.js
'use strict';

const mcpadc = require('../../');
const assert = require('assert');

let tempReading;
let tempCount = 0;
let voltageReading;
let voltageCount = 0;

const tempSensor = mcpadc.open(5, {speedHz: 20000}, err => {
  assert(!err, 'can\'t open temp sensor');

  const next = _ => {
    tempSensor.read((err, reading) => {
      assert(!err, 'can\'t read temp sensor');
      tempReading = reading;
      tempCount += 1;
      next();
    });
  };

  next();
});

const voltageSensor = mcpadc.open(4, err => {
  assert(!err, 'can\'t open voltage sensor');

  const next = _ => {
    voltageSensor.read((err, reading) => {
      assert(!err, 'can\'t read voltage sensor');
      voltageReading = reading;
      voltageCount += 1;
      next();
    });
  };

  next();
});

setInterval(_ => {
  console.log(
    'tc: ' + tempCount +
    ', t: ' + ((tempReading.value * 3.3 - 0.5) * 100) +
    ', vc: ' + voltageCount +
    ', v: ' + (voltageReading.value * 3.3)
  );
}, 1000);



index.d.ts
/**
 * Library for interacting with MCP3XXX devices
 *
 * Supported devices:
 *
 * MCP300X: MCP3002 MCP3004 MCP3008
 *
 * MCP320X: MCP3202 MCP3204 MCP3208
 *
 * MCP330X: MCP3304
 */
declare module "mcp-spi-adc" {
  export type EightChannels = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7;

  export type FourChannels = 0 | 1 | 2 | 3;

  export type TwoChannels = 0 | 1;

  export type Channels = EightChannels | FourChannels | TwoChannels;

  /**
   * Connect to an MCP3008 ADC
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21295d.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param options Configuration values that can be overloaded
   * @param callback Called after establishing a connection to the device
   */
  export function open(
    channel: EightChannels,
    options: McpOptions<EightChannels>,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3008 ADC using the default options
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21295d.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param callback Called after establishing a connection to the device
   */
  export function open(
    channel: EightChannels,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3008 ADC
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21295d.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param options Configuration values that can be overloaded
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3008(
    channel: EightChannels,
    options: McpOptions<EightChannels>,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3008 ADC using the default options
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21295d.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3008(
    channel: EightChannels,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3004 ADC
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21295d.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param options Configuration values that can be overloaded
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3004(
    channel: FourChannels,
    options: McpOptions<FourChannels>,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3002 ADC using the default options
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21295d.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3004(
    channel: FourChannels,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3002 ADC
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21294E.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param options Configuration values that can be overloaded
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3002(
    channel: TwoChannels,
    options: McpOptions<TwoChannels>,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3002 ADC using the default options
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21294E.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3002(
    channel: TwoChannels,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3202 ADC
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21294E.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param options Configuration values that can be overloaded
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3202(
    channel: TwoChannels,
    options: McpOptions<TwoChannels>,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3202 ADC using the default options
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21294E.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3202(
    channel: TwoChannels,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3204 ADC
   * ([Datasheet]https://ww1.microchip.com/downloads/en/DeviceDoc/21298e.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param options Configuration values that can be overloaded
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3204(
    channel: FourChannels,
    options: McpOptions<FourChannels>,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3204 ADC using the default options
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21298e.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3204(
    channel: FourChannels,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3208 ADC
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21298e.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param options Configuration values that can be overloaded
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3208(
    channel: EightChannels,
    options: McpOptions<EightChannels>,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3208 ADC using the default options
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21298e.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3208(
    channel: EightChannels,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3304 ADC
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21697F.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param options Configuration values that can be overloaded
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3304(
    channel: EightChannels,
    options: McpOptions<EightChannels>,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * Connect to an MCP3304 ADC using the default options
   * ([Datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/21697F.pdf))
   *
   * @param channel Zero-indexed output channel number
   * @param callback Called after establishing a connection to the device
   */
  export function openMcp3304(
    channel: EightChannels,
    callback: (error: string) => void
  ): McpInterface;

  /**
   * SPI device surface area for interacting with the MCP3XXX device
   */
  export interface McpInterface {
    /**
     * Query the device for the current reading
     * @param callback Called after the read command has completed
     */
    read(callback: (error: string, reading: McpReading) => void): void;

    /**
     * Sever communication with the MCP3XXX device
     * @param callback Called after communication has been stopped
     */
    close(callback: (error: string) => void): void;
  }

  export interface McpReading {
    /* Device reading normalized in the range [0, 1] */
    value: number;
    /* Raw counts received from the device */
    rawValue: number;
  }

  interface McpOptions<C extends Channels> {
    /* SPI clock frequency for reaching from the channel */
    speedHz?: number;
    /* SPI bus number, e.g. /dev/spidev0.X*/
    busNumber?: number;
    /* Spi device number, e.g. /dev/spidevX.0 */
    deviceNumber?: C;
  }
}


mcp-spi-adc.js
'use strict';

const spi = require('spi-device');
const clone = require('lodash.clone');

const CONFIG_MCP3008 = Object.freeze({
  channelCount: 8,
  maxRawValue: 1023,
  defaultSpeedHz: 1350000, // See MCP3008 datasheet. 75000 * 18 = 1350000.
  transferLength: 3,
  readChannelCommand: channel =>
    Buffer.from([0x01, 0x80 + (channel << 4), 0x00]),
  rawValue: buffer => ((buffer[1] & 0x03) << 8) + buffer[2]
});

const CONFIG_MCP3004 = clone(CONFIG_MCP3008);
CONFIG_MCP3004.channelCount = 4;
Object.freeze(CONFIG_MCP3004);

const CONFIG_MCP3002 = Object.freeze({
  channelCount: 2,
  maxRawValue: 1023,
  defaultSpeedHz: 1200000, // See MCP3002 datasheet. 75000 * 16 = 1200000.
  transferLength: 2,
  readChannelCommand: channel =>
    Buffer.from([0x68 + (channel << 4), 0x00]),
  rawValue: buffer => ((buffer[0] & 0x03) << 8) + buffer[1]
});

const CONFIG_MCP3208 = Object.freeze({
  channelCount: 8,
  maxRawValue: 4095,
  defaultSpeedHz: 1000000, // See MCP3208 datasheet. 50000 * 20 = 1000000.
  transferLength: 3,
  readChannelCommand: channel =>
    Buffer.from([0x06 + (channel >> 2), (channel & 0x03) << 6, 0x00]),
  rawValue: buffer => ((buffer[1] & 0x0f) << 8) + buffer[2]
});

const CONFIG_MCP3204 = clone(CONFIG_MCP3208);
CONFIG_MCP3204.channelCount = 4;
Object.freeze(CONFIG_MCP3204);

const CONFIG_MCP3202 = Object.freeze({
  channelCount: 2,
  maxRawValue: 4095,
  defaultSpeedHz: 900000, // See MCP3202 datasheet. 50000 * 18 = 900000.
  transferLength: 3,
  readChannelCommand: channel =>
    Buffer.from([0x01, 0xa0 + (channel << 6), 0x00]),
  rawValue: buffer => ((buffer[1] & 0x0f) << 8) + buffer[2]
});

const CONFIG_MCP3201 = Object.freeze({
  channelCount: 1,
  maxRawValue: 4095,
  defaultSpeedHz: 800000, // See MCP3201 datasheet. 50000 * 16 = 800000.
  transferLength: 2,
  readChannelCommand: channel =>
    Buffer.from([0x00, 0x00]),
  rawValue: buffer => ((buffer[0] & 0x1f) << 7) + (buffer[1] >> 1)
});

const CONFIG_MCP3304 = Object.freeze({
  channelCount: 8,
  maxRawValue: 4095,
  defaultSpeedHz: 1050000, // See MCP3304 datasheet. 50000 * 21 = 1050000
  transferLength: 3,
  readChannelCommand: channel =>
    Buffer.from([0x0c + (channel >> 1), (channel & 0x01) << 7, 0x00]),
  rawValue: buffer => ((buffer[1] & 0x0f) << 8) + buffer[2]
});

class AdcChannel {
  constructor(config, channel, options, cb) {
    if (typeof options === 'function') {
      cb = options;
      options = {};
    }

    if (!Number.isInteger(channel) ||
        channel < 0 ||
        channel > config.channelCount - 1) {
      throw RangeError(
        '\'' + channel + '\'' + ' is not a valid channel number'
      );
    }

    this._config = config;
    this._readChannelCommand = config.readChannelCommand(channel);
    this._speedHz = options.speedHz || config.defaultSpeedHz;
    this._device = spi.open(
      options.busNumber || 0,
      options.deviceNumber || 0,
      cb
    );

    return this;
  }

  read(cb) {
    const message = [{
      byteLength: this._config.transferLength,
      sendBuffer: this._readChannelCommand,
      receiveBuffer: Buffer.alloc(this._config.transferLength),
      speedHz: this._speedHz
    }];

    this._device.transfer(message, (err, message) => {
      const reading = {};

      if (err) {
        return cb(err);
      }

      reading.rawValue = this._config.rawValue(message[0].receiveBuffer);
      reading.value = reading.rawValue / this._config.maxRawValue;

      cb(null, reading);
    });

    return this;
  }

  close(cb) {
    this._device.close(cb);
    return null;
  }
}

module.exports.openMcp3002 = (channel, options, cb) =>
  new AdcChannel(CONFIG_MCP3002, channel, options, cb);

module.exports.openMcp3004 = (channel, options, cb) =>
  new AdcChannel(CONFIG_MCP3004, channel, options, cb);

module.exports.openMcp3008 = (channel, options, cb) =>
  new AdcChannel(CONFIG_MCP3008, channel, options, cb);
module.exports.open = module.exports.openMcp3008;

module.exports.openMcp3201 = (channel, options, cb) =>
  new AdcChannel(CONFIG_MCP3201, channel, options, cb);

module.exports.openMcp3202 = (channel, options, cb) =>
  new AdcChannel(CONFIG_MCP3202, channel, options, cb);

module.exports.openMcp3204 = (channel, options, cb) =>
  new AdcChannel(CONFIG_MCP3204, channel, options, cb);

module.exports.openMcp3208 = (channel, options, cb) =>
  new AdcChannel(CONFIG_MCP3208, channel, options, cb);

module.exports.openMcp3304 = (channel, options, cb) =>
  new AdcChannel(CONFIG_MCP3304, channel, options, cb);

